File: exercise/java/day22/src/main/java/eid/Year.java
package eid;

import io.vavr.control.Either;
import lombok.EqualsAndHashCode;
import lombok.experimental.ExtensionMethod;

@ExtensionMethod(StringExtensions.class)
@EqualsAndHashCode
public class Year {
    private final int value;

    private Year(int value) {
        this.value = value;
    }

    public static Year fromInt(int potentialYear) {
        return new Year(potentialYear);
    }

    public static Either<ParsingError, Year> parseYear(String potentialYear) {
        return potentialYear
                .toInt()
                .filter(x -> x >= 0 && x <= 99)
                .map(Year::new)
                .toEither(new ParsingError("year should be between 0 and 99"));
    }

    @Override
    public String toString() {
        return String.format("%02d", value);
    }
}

File: exercise/python/day22/eid/year.py
from typing import Union

from eid.parsing import parse_or_error
from eid.parsing_error import ParsingError

PARSING_ERROR_MESSAGE = "Year should be positive and lt 100"


def is_valid(value):
    return 0 <= value <= 99


class Year:
    def __init__(self, value: int):
        if not is_valid(value):
            raise ValueError(PARSING_ERROR_MESSAGE)
        self.value = value

    def __str__(self):
        return f"{self.value:02}"


def parse_year(text: str) -> Union[ParsingError, Year]:
    return parse_or_error(text, lambda x: is_valid(x), Year, PARSING_ERROR_MESSAGE)
